syntax = "proto3";

service PaymentGateway {
  rpc Login(LoginRequest) returns (LoginResponse);     
  rpc ProcessPayment(PaymentRequest) returns (PaymentResponse); 
  rpc GetBalance(GatewayBalanceRequest) returns (GatewayBalanceResponse);
  rpc Ping(PingRequest) returns (PingResponse); 
}

service Bank {
  rpc Prepare(PrepareRequest) returns (Vote);         
  rpc Commit(CommitRequest) returns (Ack);            
  rpc Abort(AbortRequest) returns (Ack);
  rpc GetBalance(BankBalanceRequest) returns (BankBalanceResponse); 
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  string token = 1;
}

message PaymentRequest {
  string transaction_id = 1;
  string from_account = 2;
  string to_account = 3;
  double amount = 4;
}

message PaymentResponse {
  bool success = 1;
}

message PrepareRequest {
  string transaction_id = 1;
  string from_account = 2;
  string to_account = 3;
  double amount = 4;
}

message Vote {
  bool vote = 1;
}

message CommitRequest {
  string transaction_id = 1;
}

message AbortRequest {
  string transaction_id = 1;
}

message Ack {
  bool success = 1;
}

// Separate messages for PaymentGateway and Bank
message GatewayBalanceRequest {
  string username = 1; // Username to fetch balances for
}

message GatewayBalanceResponse {
  map<string, double> accounts = 1; // Map of account numbers to balances
}

message BankBalanceRequest {
  string account_number = 1; // Account number to fetch balance for
}

message BankBalanceResponse {
  double balance = 1; // Balance of the account
}

// New messages for Ping
message PingRequest {
  string message = 1; // Optional message for debugging
}

message PingResponse {
  string message = 1; // Response message
}